### PR TITLE
Autocreate parallel fieldprops and use compressed ones in initstateequil.hh

### PR DESCRIPTION
**Pull request by blattms**
With this PR `ParallelEclipseState` will be able to autocreate data for somekeywords that have not been explicitly specified in the input file (just like the serial version). For now this still happens by wasting some space, creating global arrays of cartesian size and then compressing them manually. We use this to get rid off using the get_global_* accessors initstateequil.hh except for the temperature stuff which appears to be buggy in parallel (issue #2423) and hence might lead to differing results.

As a drive-by improvement we get rid off (at least) one unneccessary copy (PVTNUM) and on my system this give us  a 10% performance boost for Norne on 4 processors due to faster assembly.

### CODE REVIEW

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
BTW: temperature fixes will come in  a different PR.

**Comment by blattms** [comment]
benchmark

File: opm/simulators/utils/ParallelEclipseState.cpp [hunk 1]
```diff
@@ -98,8 +113,23 @@ const std::vector<double>& ParallelFieldPropsManager::get_double(const std::stri
 std::vector<double> ParallelFieldPropsManager::get_global_double(const std::string& keyword) const
 {
     std::vector<double> result;
+    int exceptionThrown{};
+
     if (m_comm.rank() == 0)
```
Comment (akva2) [inline opm/simulators/utils/ParallelEclipseState.cpp]: add braces around multi-line if's please.

File: opm/simulators/utils/ParallelEclipseState.cpp [hunk 2]
```diff
@@ -66,7 +66,19 @@ const std::vector<int>& ParallelFieldPropsManager::get_int(const std::string& ke
 {
     auto it = m_intProps.find(keyword);
     if (it == m_intProps.end())
-        OPM_THROW(std::runtime_error, "No integer property field: " + keyword);
+    {
+        // Some of the keywords might are defaulted.
```
Comment (akva2) [inline opm/simulators/utils/ParallelEclipseState.cpp]: might be

File: opm/simulators/utils/ParallelEclipseState.cpp [hunk 3]
```diff
@@ -66,7 +66,19 @@ const std::vector<int>& ParallelFieldPropsManager::get_int(const std::string& ke
...
+        // We will let rank 0 create them and distribut them using get_global_int
```
Comment (akva2) [inline opm/simulators/utils/ParallelEclipseState.cpp]: distributE

File: opm/simulators/utils/ParallelEclipseState.cpp [hunk 4]
```diff
@@ -104,7 +116,18 @@ const std::vector<double>& ParallelFieldPropsManager::get_double(const std::stri
 {
     auto it = m_doubleProps.find(keyword);
     if (it == m_doubleProps.end())
-        OPM_THROW(std::runtime_error, "No double property field: " + keyword);
+    {
+        // Some of the keywords might are defaulted.
+        // We will let rank 0 create them and distribut them using get_global_int
```
Comment (akva2) [inline opm/simulators/utils/ParallelEclipseState.cpp]: distributE

**Comment by akva2** [comment]
small stuff, otherwise it looks like an improvement.

**Comment by alfbr** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                              | Relative |
| ------- | ------------------------------------------ | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1 | 1.051    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8 | 1.001    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1 | 0.984    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8 | 0.95     |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=678

**Comment by blattms** [comment]
O my ... I blame my weired aura for the failures.

For flow we externally set eclState to be ParallelEclipseState and hence the casts to it in `EclCpGridVanguard` are safe. I guess for the failing tests equil and ecl_output we actually do not do that and end up with a real `EclipseState` base class and now cast this to something else. Not sure how to fix that.

**Comment by akva2** [comment]
Either you do the weak condition I used earlier (mpiSize > 1), or you use a dynamic_cast.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
force pushed again as indentation was messed up.

**Comment by blattms** [comment]
enkins build this please

File: ebos/eclcpgridvanguard.hh [hunk 5]
```diff
@@ -189,17 +191,29 @@ public:
...
+                    }
 
-                PropsCentroidsDataHandle<Dune::CpGrid> handle(*grid_, eclState, eclGrid, this->centroids_,
-                                                              cartesianIndexMapper());
-                defunctWellNames_ = std::get<1>(grid_->loadBalance(handle, edgeWeightsMethod, &wells, faceTrans.data()));
+                    PropsCentroidsDataHandle<Dune::CpGrid> handle(*grid_, eclState, eclGrid, this->centroids_,
+                                                                  cartesianIndexMapper());
+                    defunctWellNames_ = std::get<1>(grid_->loadBalance(handle, edgeWeightsMethod, &wells, faceTrans.data()));
+                }
+                catch(std::bad_cast e)
```
Comment (akva2) [inline ebos/eclcpgridvanguard.hh]: catch by ref

File: ebos/eclcpgridvanguard.hh [hunk 6]
```diff
@@ -218,12 +232,24 @@ public:
...
         if (mpiSize > 1) {
-            static_cast<ParallelEclipseState&>(this->eclState()).switchToDistributedProps();
+            try
+            {
+                auto& parallelEclState = dynamic_cast<ParallelEclipseState&>(this->eclState());
+                // reset cartesian index mapper for auto creation of field properties
+                parallelEclState.resetCartesianMapper(cartesianIndexMapper_.get());
+                parallelEclState.switchToDistributedProps();
+            }
+            catch(std::bad_cast e)
```
Comment (akva2) [inline ebos/eclcpgridvanguard.hh]: catch by ref.

File: ebos/eclcpgridvanguard.hh (hunk 5 above)
Comment (blattms) [reply ebos/eclcpgridvanguard.hh]: mmh. I know "catch by reference and throw by value". To satisfy both I would need to make a copy anyway. Shouldn't the copy in the catch clause be elided anyway (probably still not guaranteed)?

Summoning @bska.

Comment (akva2) [reply ebos/eclcpgridvanguard.hh]: thing is std::bad_cast is polymorphic, so you get a warning, https://ci.opm-project.org/job/opm-simulators-PR-builder/746/gcc/new/source.9c010f09-9839-4bb1-807a-2db5201de9aa/#245

Comment (blattms) [reply ebos/eclcpgridvanguard.hh]: ok, I am convinced.

**Comment by blattms** [comment]
jenkins run norne norne_parallel please

**Pull request merged**